setversion.py

```diff
@@ -0,0 +1,120 @@
+        datestr = date.today().strftime("%B %d, %Y")
+        cl1 = "CbAPI {0} - Released {1}".format(ctxt["version"], datestr)
+        cl2 = "".ljust(len(cl1), "-")
+        updates = "\n\nUpdates\n\n.. (add your updates here)\n"
+        return line + "\n" + cl1 + "\n" + cl2 + updates
+    return None
+
+
+def doc_conf_rewriter(line, ctxt):
+    vexpr = ctxt.get("doc_version_expr", None)
+    if not vexpr:
+        vexpr = re.compile(r"^version = ")
+        ctxt["doc_version_expr"] = vexpr
+    rexpr = ctxt.get("doc_release_expr", None)
+    if not rexpr:
+        rexpr = re.compile(r"^release = ")
+        ctxt["doc_release_expr"] = rexpr
+    if vexpr.match(line):
+        t = re.match(r"^(\d+\.\d+)\.", ctxt["version"])
+        vn = ctxt["version"]
+        if t:
+            vn = t.group(1)
+        return "version = u'{0}'\n".format(vn)
+    if rexpr.match(line):
+        return "release = u'{0}'\n".format(ctxt["version"])
+    return None
+
+
+def setup_rewriter(line, ctxt):
+    expr = ctxt.get("setup_expr", None)
+    if not expr:
+        expr = re.compile(r"^(\s*)version=")
+        ctxt["setup_expr"] = expr
+    m = expr.match(line)
+    if m:
+        return "{0}version='{1}'\n".format(m.group(1), ctxt["version"])
+    return None
+
+
+def init_rewriter(line, ctxt):
+    expr = ctxt.get("init_expr", None)
+    if not expr:
+        expr = re.compile(r"^__version__ = ")
+        ctxt["init_expr"] = expr
+    if expr.match(line):
+        return "__version__ = '{0}'\n".format(ctxt["version"])
+    return None
+
+
+def rewrite_file(infilename, rewritefunc, ctxt):
+    outfilename = infilename + ".new"
+    infile = open(infilename, "r")
+    outfile = open(outfilename, "w")
+    try:
+        s = infile.readline()
+        while s:
+            s2 = rewritefunc(s, ctxt)
+            if s2:
+                outfile.write(s2)
+            else:
+                outfile.write(s)
+            s = infile.readline()
+    finally:
+        infile.close()
+        outfile.close()
+    if not ctxt["nodelete"]:
+        os.remove(infilename)
+        os.rename(outfilename, infilename)
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Set the version number in CbAPI source and documentation.\n"
+                                                 "Execute this on a release or hotfix branch to update "
+                                                 "the version numbers in the source.",
+                                     epilog="After running, edit docs/changelog.rst and add the new changelog "
+                                            "information under the new heading.")
+    parser.add_argument("version", help="New version number to add")
+    parser.add_argument("-n", "--nodelete", action="store_true",
+                        help="Do not delete existing files, leave new files with .new extension")
+
+    args = parser.parse_args()
+    ctxt = {"version": args.version, "nodelete": args.nodelete}
```

Comment:
Could we add some validation on the version string?

Reply:
We could, if we knew what to validate it against. Suggestions?

Reply:
#.#.# would be my expected matching since that is the version format we are following

Reply:
That can be done, but do we want to leave the possibility of labeling "beta" releases, such as "1.6.0b1"?

Reply:
It's only going to validate as three numeric values separated by dots.